### ISSUE TITLE
Define class types, attributes, and methods

### ISSUE THREAD

**Issue by LiamDBailey** [LiamDBailey/ecogon#1]
## Tile (Parent object)
*Attributes*:
- Neighbours: List of neighbouring objects of class tile. Length of list will define the number of sides used for polygon rendering.

*Methods*:
None

---
## Habitat (Child of tile class)
*Attributes*:
- HabitatType: String. Forest, Forest Edge, Grassland, Lake etc.

*Methods*:
- CheckNeighbours: Check that all neighbours are of compatible HabitatType (e.g. Forest cannot border Grassland).

---
## Plant (Child of habitat class)
All organisms also have a habitat type, so they can be child of habitat class

*Attributes*:
- PlantType: String. Flower, Fruit, Tree etc.

*Methods*:
- CheckHabitat: Check that at least one neighbour is of class Habitat and HabitatType of self.

---
## Animal (Child of habitat class)
All organisms also have a habitat type, so they can be child of habitat class

*Attributes*:
- AnimalType: String. Terrestrial, Flying etc.
- Size: Integer. 1-5
- Score: Integer. Points gained from completing.
- Requirements: List (?) of required neighbours to complete the tile.
- Complete: Logical. Have requirements been met?

*Methods*:
- CheckResources: Check whether neighbours meet requirements. If yes, Complete = TRUE.

---
## Player

*Attributes*:
- Cards: List of objects of class tile. Cards that the player can choose to play during a turn.
- HandSize: Integer. Maximum size of 'Cards' attribute.
- Turn: Logical. Is it this players turn?

*Methods*:
- PlaceCard: Move a card for 'Cards' attribute into the 'Neighbours' attribute of a tile.
- CheckHandSize: Compare length of 'Cards' with 'HandSize'.

---
## Event card

*Attributes*:
- TargetClass: String. Habitat, Plant, Animal.
- TargetPlantType: Flower, Fruit etc.
- TargetAnimalType: Terrestrial, Water etc.
- TargetAnimalSize: Integer. 1-5.
- NumberTargets: Integer. Number of card targets that must be selected.
- Targets: List of objects of tile class. Cards that will be affected by methods.

*Methods*:
- RemoveCard: Remove objects from 'Targets' attribute from the game.
- SwapCard: Change the place of first and second card in 'Targets' attribute.

**Comment by LiamDBailey** [comment]
First go at classes, attributes and methods. It may be better to have an 'Organism' class, rather than separate 'Plant' and 'Animal' classes. That will reduce the number of attributes needed to define event card targets (e.g. no AnimalType v. PlantType). An object of class 'Organism' that is a plant will have size 0, score 0, and requirements NA.

**Comment by LiamDBailey** [comment]
## Game
*Attributes*:
- TileDeck: List of objects of tile class. Remaining tiles that can be picked up by players.
- TileDiscard: List of objects of tile class. Tiles that have been discarded (*check if this is ever needed*)
- EventDeck: List of objects of event card class. Remaining event cards that can occur.
- EventDiscard: List of objects of event card class. Events that have already occurred (*Used in one event card case. Maybe easier to ignore this case to begin with, for simplicity.*)
- Players: List of objects of class player. All players in the game.
- EventCardTurns: Integer. Number of turns until the next event card is drawn.

*Methods*:
- PlayEvent: Check EventCardTurns attribute and run event card method if EventCardTurns = 0.
- CheckWin: End game and check win condition if length of either TileDeck or EventDeck are 0.

**Comment by aMollusk** [comment]
**Tile:**

Will probs need an `addNeighbour()` and `removeNeighbour()` method.

**Plants etc:**

I'm not sure inheriting from habitat is correct here? Do all organisms require the methods that habitats have? If so, this is fine.

RE:
> First go at classes, attributes and methods. It may be better to have an 'Organism' class, rather than separate 'Plant' and 'Animal' classes. That will reduce the number of attributes needed to define event card targets (e.g. no AnimalType v. PlantType). An object of class 'Organism' that is a plant will have size 0, score 0, and requirements NA.

This is probably a personal choice. It probably comes down to whether or not these will require different methods. Happy to stick with organism if they all have the same possible functionality. It's still worth having `organisimType: OrganismTypes.Animal` attributes though. It depends on whether we want to be checking logic against class instance types or attribute types.

I think it comes down to functionality differences.

```
// some function has accepted a mixed array of Animals or Organisms

for(...) {
  const organism = listOfThings[index];
  organism.run() // This causes an exception because plants don't have a method of run
  
  // so we'd be forced to write
  if(organism instanceof Animal) {
    organism.run()
  }
  
  // The above is normal, but the following would be bad (imo)
  if(organism.organismType === OrganismTypes.Animal) {
    organism.run()
  }
}
```

Bottom line is that I think it's chill to simplify in to one class as long as we don't see them having a bunch of different methods.

Only thing I think is missing is a board, i.e. what's been already played.

**Comment by aMollusk** [comment]
The more I think about it the more it seems a bit strange that organisms inherit from habitat, especially as organisms belong to a habitat.

Would we need to track all organisms that are _currently_ associated with a habitat?

**Comment by LiamDBailey** [comment]
So remember that each organism will also have a habitat type (i.e. habitat where it can occur). From the rules:

> At any time during the game when you play your habitat card, please remember that a forest and a meadow [grassland] card should never lie next to each other. This is also true for all other types of cards *including animal and plant cards*. Only the forest edge can host both forest and meadow [grassland] species.

This also applies to the water and shoreline habitats. I think the easiest way to think about it is with a fish. You shouldn't find a fish next to a forest (or a forest animal). It should only be near a water or shoreline. So that means the `CheckNeighbours` method will be relevant for *any* tile.

RE: Plant/Animal v. Organism.

Plants will have very few methods. All they need to do is check whether they are in the right habitat. Animals will have at least a 'CheckResources' method that will not be relevant for plants. Animals will also have lots of extra attributes (e.g. a size, a type [e.g. terrestrial, flying]). So, if we created a single Organism class we would be doing the 'bad' logical check you show above. So in that case, I think it's better to keep the animal and plant classes separate.

**Comment by aMollusk** [comment]
I wonder if in that case we add `checkNeighbour` to Tile (as an interface that has to be implemented by child classes), and we make `Organism` extend `Tile`, and make `Habitat` also extend `Tile`. 

Without getting too deep in to biology, an animal isn't really a child of a habitat, in the same way that a truck isn't a child of a highway - it's more of a descendant of a vehicle. It's kind of trivial here, but an example where it breaks down is that maybe habitat has `rain()`... a bird can't `.rain()`, Liam.... that's silly.

p.s. this is largely why people don't like inheritance... because as you can see, it can breakdown really easily :P

**Comment by aMollusk** [comment]
I'll point out that the `rain()` example might not really apply here, because we're building a game, not an environment

**Comment by LiamDBailey** [comment]
Having `Habitat` and `Organism` inherit from `Tile` makes more logical sense to me. tbh, the scientist in me likes the hierarchical classification of class inheritance 😆.

I think for now `Habitat` will have no additional methods or attributes to those of `Tile`, but maybe good to make the organisation logical now in case we want to add additional functionality later.
